### PR TITLE
Refactor builtin macro

### PR DESCRIPTION
**Pull request by edwin0cheng**
Refactor builtin macro and add some local tests.

### CODE REVIEW

**Comment by DJMcNab** [comment]
Can you rename `BuiltinExpander` to `BuiltinFnLikeExpander`?

We will want to support built in attribute and derive style macros, and I need this so that I can start adding procedural macro support fairly easily.

**Comment by edwin0cheng** [comment]
@DJMcNab I don't have strong opinion about the naming :)

**Comment by DJMcNab** [comment]
Thanks!

File: crates/ra_hir_expand/src/builtin_macro.rs [hunk 1]
```diff
@@ -9,26 +9,52 @@ use crate::{
 use crate::quote;
 
 #[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
-pub enum BuiltinExpander {
+pub enum BuiltinFnLikeExpander {
```
Comment (DJMcNab) [inline crates/ra_hir_expand/src/builtin_macro.rs]: :+1: 

Can this be generated by `register_builtin`? If we did that we could use `kind as u16` to index into `BUILTIN_MACROS` (with a `[debug_]assert` that `result.kind==kind`). Ideally we would use an array (using `BuiltinFnLikeExpander::$last as usize` for the array size)

I suppose it probably doesn't matter - it depends on how the manual iteration is optimised. I just like looking for some stupid optimisation opportunities.

Comment (edwin0cheng) [reply crates/ra_hir_expand/src/builtin_macro.rs]: Good idea, I just use macro for all the setup and eliminate the immediate struct too.

**Comment by matklad** [comment]
bors r+

I like the test setup!

**Pull request merged**